crates/binstalk/src/helpers/git.rs

```diff
@@ -0,0 +1,80 @@
+use std::{num::NonZeroU32, path::Path, str::FromStr, sync::atomic::AtomicBool};
+
+use gix::{clone, create, open, progress::Discard, remote, Url};
+use thiserror::Error as ThisError;
+use tracing::debug;
+
+#[derive(Debug, ThisError)]
+#[non_exhaustive]
+pub enum GitError {
+    #[error("Failed to prepare for fetch: {0}")]
+    PrepareFetchError(#[source] Box<clone::Error>),
+
+    #[error("Failed to fetch: {0}")]
+    FetchError(#[source] Box<clone::fetch::Error>),
+
+    #[error("Failed to checkout: {0}")]
+    CheckOutError(#[source] Box<clone::checkout::main_worktree::Error>),
+}
+
+impl From<clone::Error> for GitError {
+    fn from(e: clone::Error) -> Self {
+        Self::PrepareFetchError(Box::new(e))
+    }
+}
+
+impl From<clone::fetch::Error> for GitError {
+    fn from(e: clone::fetch::Error) -> Self {
+        Self::FetchError(Box::new(e))
+    }
+}
+
+impl From<clone::checkout::main_worktree::Error> for GitError {
+    fn from(e: clone::checkout::main_worktree::Error) -> Self {
+        Self::CheckOutError(Box::new(e))
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct GitUrl(Url);
+
+impl FromStr for GitUrl {
+    type Err = gix::url::parse::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Url::try_from(s).map(Self)
+    }
+}
+
+#[derive(Debug)]
+pub struct Repository(gix::Repository);
+
+impl Repository {
+    /// WARNING: This is a blocking operation, if you want to use it in
```

Comment:
Is there no async variant from gix? For a mostly-network op that seems odd

Reply:
No, I remembered that @Byron explicitly said it's not a goal of `gix`.
I also checked the API of `gix`, it seems that if we want to use the `async` version of it then we won't have access to easy-to-use high-level APIs.

Reply:
Fair enough